tooi/api/timeline.py

```diff
@@ -60,15 +62,115 @@ async def fetch_timeline(
 
 class Timeline(ABC):
     """
-    Base class for a timeline.  This provides some useful generators that subclasses can use.
+    Base class for a timeline.  A timeline generates Events, which can be retrieved via its fetch()
+    or fetch_wait() methods.  The type of events the timeline generates depends on which subclass is
+    instantiated.
     """
 
-    def __init__(self, name, instance: InstanceInfo):
+    # How many events to queue at once before attempts to queue more events start to block.
+    # 128 should be a reasonable amount to allow a large timeline refresh to queue fully
+    # without blocking.  Increasing this doesn't significantly affect memory usage, because these
+    # events will be turned into timeline events anyway, but it may cause the timeline to be updated
+    # slower if many events being added block get_events().  This may need to be tuned later or
+    # perhaps removed entirely (set to zero).
+    QUEUE_SIZE = 128
+
+    def __init__(self, name, instance: InstanceInfo, can_update: bool = False):
         self.instance = instance
         self.name = name
+        self.can_update = can_update
+        self._update_running = AsyncAtomic[bool](False)
+        self._update_task = None
+        self._periodic_refresh_task = None
+        self._queue = asyncio.Queue()
+
+    def _assert_can_update(self):
+        if not self.can_update:
+            raise (NotImplementedError("this Timeline cannot update"))
+
+    def close(self):
+        if (update_task := self._update_task) is not None:
+            update_task.cancel()
+
+        if (periodic_refresh_task := self._periodic_refresh_task) is not None:
+            periodic_refresh_task.cancel()
+
+    async def update(self):
+        self._assert_can_update()
+        await self._interlocked_update()
+
+    def periodic_refresh(self, frequency: int):
+        self._assert_can_update()
+        self._periodic_refresh_task = run_async_task(self._periodic_refresh(frequency))
+
+    async def _periodic_refresh(self, frequency: int):
+        while True:
+            await asyncio.sleep(frequency)
+            await self._interlocked_update()
+
+    async def get_events(self) -> list[Event]:
+        """
+        Return a list of pending events which have been queued to the timeline since the last time
+        get_events() or get_events_wait() were called.  This function always returns immediately.
+        If no events are available, an empty list will be returned.
+        """
+        events = []
+
+        while True:
+            try:
+                event = self._queue.get_nowait()
+                self._queue.task_done()
+            except asyncio.QueueEmpty:
+                return events
+
+            events.append(event)
+
+        return events
```

Comment:
This line of code is not reachable.

Reply:
fixed in cf6da1e45620a81f5884b96618e283c2fd2fab96